### PR TITLE
IGNITE-18521 Sql. Query hangs if exception is thrown in mapping phase.

### PR DESCRIPTION
**Pull request by xtern**

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 1]
```diff
@@ -576,36 +572,48 @@ private void submitFragment(String initiatorNode, String fragmentString, Fragmen
...
 
-                        for (String nodeName : fragmentDesc.nodeNames()) {
-                            sendFragment(nodeName, fragment, fragmentDesc);
+                            break;
                         }
                     }
-                } catch (Throwable e) {
-                    root.thenAccept(root -> root.onError(e));
+                } catch (Throwable t) {
+                    LOG.error("An error occurred while executing the query.", t);
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:614]: we should not use ERROR severity here. PLease take a look at [this](https://cwiki.apache.org/confluence/display/IGNITE/Java+Logging+Guidelines#JavaLoggingGuidelines-ERROR) guidelines

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 2]
```diff
@@ -576,36 +572,48 @@ private void submitFragment(String initiatorNode, String fragmentString, Fragmen
...
+
+                    root.completeExceptionally(t);
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:616]: what if `root` has been already completed at this point?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java (hunk 1 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java]: thanks, additional logging removed

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java (hunk 2 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java]: Fixed a bit to handle such scenario :thinking:

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java [hunk 3]
```diff
@@ -607,7 +633,8 @@ protected TestTable createTable(String name, int size, IgniteDistribution distr,
                 new TableDescriptorImpl(columns, distr),
                 name,
                 ColocationGroup.forNodes(nodeNames),
-                size
+                size,
+                () -> mappingException
```
Comment (AMashenkov) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java:637]:
```suggestion
                size
```
Let's just override the `colocationGroup` method instead, and add a short explanation on why we throws an exception from here.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/framework/TestTable.java [hunk 4]
```diff
@@ -234,6 +239,12 @@ public boolean rolledUpColumnValidInsideAgg(
     /** {@inheritDoc} */
     @Override
     public ColocationGroup colocationGroup(MappingQueryContext ctx) {
+        RuntimeException ex = mappingExceptionSupplier.get();
+
+        if (ex != null) {
+            throw ex;
+        }
+
```
Comment (AMashenkov) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/framework/TestTable.java:247]:
```suggestion
```
This looks like a dirty hack.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java (hunk 3 above)
Comment (xtern) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java]: Thanks, I've fixed it as per your suggestion.

**Review by AMashenkov — APPROVED** [review]

**Pull request merged**